osh/builtin.py

```diff
@@ -512,12 +513,25 @@ def __call__(self, arg_vec):
 
     names = arg_vec.strs[i:]
     if arg.n is not None:  # read a certain number of bytes
+      stdin = sys.stdin.fileno()
       try:
         name = names[0]
       except IndexError:
         name = 'REPLY'  # default variable name
-      s = posix.read(sys.stdin.fileno(), arg.n)
-      #log('read -n: %s = %s', name, s)
+      orig_attrs = termios.tcgetattr(stdin)
+      attrs = termios.tcgetattr(stdin)
+      # disable canonical (buffered) mode
+      # see `man termios` for an extended discussion
```

Comment:
My understanding  is a little fuzzy here, but I don't think this makes sense because stdin isn't always a terminal.

Well first I'm not clear exactly what problem this is solving.  Do we have a test that fails with the old code and passes with this code (whether manual or automated)?

Is it that you want `read -n 1` to return after a keypress?  OK I see that doesn't work in OSH.

Why doesn't it work?  I don't think Our posix.read(0, 1) call is supposed to buffer anything?  If I look at the source to `native/posixmodule.c`, it calls `read()` directly.  So is the source of buffering in libc?

But what happens when stdin is not a terminal?  e.g. a file?   `read -n 1 < myfile`

I looked at the source to bash and dash.  It looks like they both call `read(1)` (a single byte) ALL the time.   They don't appear to do anything with the terminal settings ?

Anyway I guess I'm interested in an analysis of what doesn't work about the current code, and what other shells do.  It does seem like something is wrong, but I don't know what.

Reply:
Hm Ok I missed the strace... let me look at that.  I was looking at the bash source but didn't see it doing anything like that.

Reply:
OK now I see in `builtins/read.def`, bash does `ttfd_onechar()`, which presumably does something just like this.

dash doesn't implement `read -n`, so it doesn't apply.

OK I will look at this in more detail when I get back.  Thanks for the patch.

Reply:
That's a good point about non-interactive stdin, I'll add some error handling for that case.

Reply:
The buffering I think is happening below libc in the kernel drivers for the terminal. I don't have an automated test case, but if you do `read -n 2` it should read two characters _before_ you hit newline, which it wasn't doing before.